zp-core/zp-extensions/downloadList.php

```diff
@@ -284,15 +284,66 @@ static function noFile() {
 		} else {
 			$file = basename($_zp_downloadfile);
 		}
+
+		$back_url = preg_replace('/[&|?]download=.+/', '', $_SERVER["REQUEST_URI"]);
+
+		header('Content-Type: text/html; charset=' . LOCAL_CHARSET);
+		header("HTTP/1.0 404 Not Found");
+		header("Status: 404 Not Found");
+		zp_apply_filter('theme_headers');
 		?>
-		<script type="text/javascript">
-			// <!-- <![CDATA[
-			window.onload = function () {
-				alert('<?php printf(gettext('File “%s” was not found.'), $file); ?>');
-			}
-			// ]]> -->
-		</script>
+		<!DOCTYPE html>
+		<html<?php printLangAttribute(); ?>>
+			<head>
+				<meta name="viewport" content="width=device-width, initial-scale=1.0">
+				<meta charset="<?php echo LOCAL_CHARSET; ?>">
+				<meta name="ROBOTS" content="NOINDEX, FOLLOW">
+				<title><?php echo gettext("Object not found") . ': ' . $file . ' | ' . html_encode(getBareGalleryTitle()) ?></title>
+				<style>
+				body {
+					background: white;
+				}
+				div {
+					text-align: center;
+					background: url(<?php echo WEBPATH . '/' . ZENFOLDER ?>/images/zen-logo.png) 50% calc(30vh - 70px) no-repeat;
+					padding: 30vh 1em 0;
+					color: #f56a6a;
+					text-transform: uppercase;
+				}
+				a {
+					color: #718086;
+					margin: 0 .5em;
+				}
+				</style>
+			</head>
+			<body>
+				<div>
+					<?php $file = sprintf(gettext('File “%s” was not found.'), $file); ?>
+					<h3><?php echo $file ?></h3>
+					<a href="<?php echo $back_url; ?>"><?php echo gettext('Back') ?></a>
+					<a href="<?php echo WEBPATH; ?>/"><?php echo gettext('Home') ?></a>
+				</div>
+			</body>
+		</html>
 		<?php
+		exitZP();
+	}
+
+	static function checkAccess(&$hint = NULL, &$show = NULL) {
+		$hash = getOption('downloadList_password');
+		if (GALLERY_SECURITY != 'public' || $hash) {
+			//	credentials required to download
+			if (!zp_loggedin((getOption('downloadList_rights')) ? FILES_RIGHTS : ALL_RIGHTS)) {
+				$user = getOption('downloadList_user');
+				zp_handle_password('zpcms_auth_download', $hash, $user);
+				if ((!empty($hash) && zp_getCookie('zpcms_auth_download') != $hash)) {
+					$show = (!empty($user));
+					$hint = get_language_string(getOption('downloadList_hint'));
+					return false;
+				}
+			}
+		}
+		return true;
```

Comment:
Handling of download credentials is in now in a separate method, as before $hint and $show scope was global because they were declared outside any function.